test/e2e/e2e_test.go

```diff
@@ -51,23 +56,169 @@ var _ = Describe("leaderWorkerSet e2e tests", func() {
+		lws = testing.BuildLeaderWorkerSet(ns.Name).Replica(4).Size(5).RestartPolicy(v1.RecreateGroupOnPodRestart).Obj()
+		testing.CreateLws(ctx, k8sClient, lws)
+
+		Expect(*lws.Spec.Replicas).To(Equal(int32(4)))
+		Expect(*lws.Spec.LeaderWorkerTemplate.Size).To(Equal(int32(5)))
+		Expect(lws.Spec.LeaderWorkerTemplate.RestartPolicy).To(Equal(v1.RecreateGroupOnPodRestart))
 	})
 
-	It("Can rolling update lws", func() {
+	It("Can perform a rolling update", func() {
 		lws := testing.BuildLeaderWorkerSet(ns.Name).Obj()
-		Expect(k8sClient.Create(ctx, lws)).To(Succeed())
+		testing.CreateLws(ctx, k8sClient, lws)
 
-		// Wait for leaderWorkerSet ready then update it.
+		// Wait for leaderWorkerSet to be ready then update it.
 		testing.ExpectLeaderWorkerSetAvailable(ctx, k8sClient, lws, "All replicas are ready")
 		testing.UpdateWorkerTemplate(ctx, k8sClient, lws)
 
 		testing.ExpectValidLeaderStatefulSet(ctx, lws, k8sClient)
 		testing.ExpectValidWorkerStatefulSets(ctx, lws, k8sClient, true)
 		testing.ExpectValidPods(ctx, k8sClient, lws)
-		// Wait for leaderWorkerSet ready again.
+		// Wait for leaderWorkerSet to be ready again.
 		testing.ExpectLeaderWorkerSetAvailable(ctx, k8sClient, lws, "All replicas are ready")
 	})
+
+	It("Adds env vars to containers when using TPU", func() {
+		leaderPodSpec := testing.MakeLeaderPodSpecWithTPUResource()
+		workerPodSpec := testutils.MakeWorkerPodSpecWithTPUResource()
+		lws = testing.BuildLeaderWorkerSet(ns.Name).Replica(2).Size(2).LeaderTemplateSpec(leaderPodSpec).WorkerTemplateSpec(workerPodSpec).Obj()
+		testing.CreateLws(ctx, k8sClient, lws)
+
+		lwsPods := &corev1.PodList{}
+		testing.GetPods(ctx, lws, k8sClient, lwsPods)
+
+		Eventually(func() (bool, error) {
+			var allPods corev1.PodList
+			err := k8sClient.List(ctx, &allPods, client.InNamespace(lws.Namespace))
+			if err != nil {
+				return false, err
+			}
+			for _, p := range lwsPods.Items {
+				if !testing.HasTPUEnvVarsPopulated(p) {
+					return false, nil
+				}
+			}
+			return true, nil
+		}, timeout, interval).Should(BeTrue())
+	})
+
+	It("Doesnt add env vars to containers when not using TPU", func() {
+		leaderPodSpec := testutils.MakeLeaderPodSpec()
+		workerPodSpec := testutils.MakeWorkerPodSpec()
+		lws = testing.BuildLeaderWorkerSet(ns.Name).Replica(2).Size(2).LeaderTemplateSpec(leaderPodSpec).WorkerTemplateSpec(workerPodSpec).Obj()
+
+		testing.CreateLws(ctx, k8sClient, lws)
+		lwsPods := &corev1.PodList{}
+		testing.GetPods(ctx, lws, k8sClient, lwsPods)
+
+		for _, p := range lwsPods.Items {
+			Expect(testing.HasTPUEnvVarsPopulated(p)).To(BeFalse())
+		}
+	})
+
+	It("Pod restart will not recreate the pod group when restart policy is Default", func() {
+		lws = testing.BuildLeaderWorkerSet(ns.Name).Replica(1).Size(3).RestartPolicy(v1.DefaultRestartPolicy).Obj()
+		testing.CreateLws(ctx, k8sClient, lws)
+		testing.GetPods(ctx, lws, k8sClient, &corev1.PodList{})
+
+		var leaderPod corev1.Pod
```

Comment:
I guess this is too detailed? Seems we're testing the statefulset logic here. Better to relate to `RestartPolicy`.

Reply:
+1; I don't think we want this test.